src/ClpIrV1Decoder.hpp

```diff
@@ -0,0 +1,66 @@
+#ifndef CLP_FFI_JS_CLPIRV1DECODER_HPP
+#define CLP_FFI_JS_CLPIRV1DECODER_HPP
+
+#include <emscripten/val.h>
+
+#include <clp/ffi/ir_stream/decoding_methods.hpp>
+#include <clp/ir/LogEventDeserializer.hpp>
+#include <clp/streaming_compression/zstd/Decompressor.hpp>
+
+/**
+ * Deserializes Zstd-compressed CLP IR V1 byte streams and formats extracted log events.
+ */
+class ClpIrV1Decoder {
+public:
+    /**
+     * Factory method to create a ClpIrV1Decoder instance by
+     * - Copying an input array into C++ space.
+     * - Constructing a Zstd Decompressor instance.
+     * - Constructing a LogEventDeserializer by decoding the preamble of the IR stream, which
+     * includes the encoding type.
+     * - Passing those to the ClpIrV1Decoder() for their life cycles management.
+     *
+     * @param data_array JS Uint8Array which contains the Zstd-compressed CLP IR V1 bytes.
+     * @return the created instance.
+     * @throw DecodingException if any error occurs.
+     */
+    [[nodiscard]] static auto create(emscripten::val const& data_array) -> ClpIrV1Decoder*;
+
+    // Destructor
+    ~ClpIrV1Decoder() = default;
+
+    // Explicitly disable copy and move constructor/assignment
+    ClpIrV1Decoder(ClpIrV1Decoder const&) = delete;
+    ClpIrV1Decoder& operator=(ClpIrV1Decoder const&) = delete;
+    ClpIrV1Decoder(ClpIrV1Decoder&&) = delete;
+    ClpIrV1Decoder& operator=(ClpIrV1Decoder&&) = delete;
+
+    /**
+     * Calculates the estimated number of events stored in the log.
+     *
+     * If `build_idx()` has not been called before, the function will return cFullRangeEndIdx=0,
+     * indicating that there are no events stored in the log.
+     *
+     * @return The estimated number of events in the log.
+     */
+    [[nodiscard]] auto get_estimated_num_events() -> size_t;
+
+    [[nodiscard]] auto build_idx(size_t begin_idx, size_t end_idx) -> emscripten::val;
+    [[nodiscard]] auto decode(size_t begin_idx, size_t end_idx) -> emscripten::val;
```

Comment:
Sorry for the missing background. We have the docstrings written at https://github.com/y-scope/yscope-log-viewer/blob/0d302c5ef66ea2e0d66ba1981766614f27b04225/new-log-viewer/src/typings/decoders.ts#L32 and we previously concluded that we might not want to repeat the docs if the method behaviours do not change significantly. With that said, that is in a separate repo and have caused confusions to our readers. Do you think we can put a comment to reference the above line?

Reply:
I would prefer to have the doc strings on both end. It would be more clear to C++ readers what this function is doing, as well as some C++ level assumption (like the ownership of an object) which is not proper to include in JS layer documents.

Reply:
Added docstrings for each method:
1. For the ones which didn't have any docs added, the docs were copied from the TS source.
2. For the ones which have different behaviours than the JS abstract class definitions, their docs were completed so readers do not have to always refer to the TS source.

> as well as some C++ level assumption (like the ownership of an object) which is not proper to include in JS layer documents

Are there any other specific behaviors you want to document? Feel free to make suggestions / push directly.